src/adapter/src/coord/sequencer/inner/create_materialized_view.rs

```diff
@@ -819,4 +819,50 @@ impl Coordinator {
 
         Ok(StageResult::Response(Self::send_immediate_rows(rows)))
     }
+
+    pub(crate) async fn materialized_view_explain_pushdown(
+        &mut self,
+        ctx: &ExecuteContext,
+        gid: GlobalId,
+    ) -> Result<ExecuteResponse, AdapterError> {
+        let entry = self.catalog.get_entry(&gid).clone();
+        let CatalogItem::MaterializedView(item) = &entry.item else {
+            Err(ExplainError::UnknownError(format!(
+                "wrong empty type for materialized view: {gid}"
+            )))?
+        };
+
+        let desc = self
+            .catalog
+            .try_get_physical_plan(&gid)
+            .ok_or_else(|| ExplainError::UnknownError(format!("unknown gid: {gid}")))?
+            .clone();
+
+        let id_bundle = dataflow_import_id_bundle(&desc, item.cluster_id);
+        let (_read_holds, dataflow_as_of, storage_as_of, until) =
+            self.select_timestamps(id_bundle, item.refresh_schedule.as_ref())?;
```

Comment:
AIUI, this will:
- Get a set of frontiers for this MV as though it were being scheduled now. (Which is the desired behaviour for `explain filter pushdown`: we want to know what kind of pushdown we'd see if we were re-rehydrating the view right now.)
- Take read holds on the inputs, ensuring that the call to `render_...` below will be able to snapshot at the chosen storage as-of.

I think this seems right in the case of ordinary MVs. It's not clear to me what semantics we'd want in the case of `REFRESH EVERY` or similar, however! Certainly we don't want to sit around until the next refresh interval...

(If this has an obvious answer I'll toss it in; otherwise I may just ban it for now...)

Reply:
> so usually the timestamps returned by select_timestamps won't match the as_of that we would select if we were bootstrapping the MV now.

Interesting... do you have a good intuition for _why_ these methods behave so differently? (Is it just that we have the output MV and can make a less conservative choice?

I'm a bit hesitant to just invoke the bootstrap code, since it seems to run in time proportional to all collections, where I'm interested in just the one. But I'll experiment with that and the workaround you suggest!

Reply:
> do you have a good intuition for why these methods behave so differently?

Concentrating on the most common case (write frontier for bootstrap, and `least_valid_read` for new MV):
If the MV is already there, then we most commonly just need to continue where we left off writing it. But when we are creating the MV, then we have to make a decision about what timestamp it should start from. For the latter, `least_valid_read` is a natural choice, but `least_valid_read` is usually a second or so behind the write frontier, and can be arbitrarily behind in various corner cases, as mentioned above.

> (Is it just that we have the output MV and can make a less conservative choice?

Yes, in the above common case, we can say that `bootstrap_dataflow_as_of` is less conservative, because the dataflow will start at a later time.

But generally, I wouldn't say that it's less conservative, because there are also some constraints that `bootstrap_dataflow_as_of` needs to take into account that don't come up when creating a new MV:
- Compute reconciliation (i.e., reusing already-running dataflows when envd is restarted without clusterds being restarted). This is quite tricky, and is what necessitates looking at the `StorageConstraints` (which is the thing that takes time proportional to all collections).
- A REFRESH MV might be after its last REFRESH. (We don't allow this for newly created MVs.)

Reply:
> time proportional to all collections

If you are worried about this, then how about adding a histogram metric for how much time this takes? Then we could observe how it actually looks in practice. And then if it turns out to be actually too slow, then we'll figure something out. (For example, we could write a version of `collect_dataflow_storage_constraints` that looks at not all collections but only at the ones relevant to one EXPLAIN command.)

Btw. there are things that we do every second that take time proportional to the number of collections:
- `advance_timelines` calls the rather complicated `update_read_holds` -> `set_read_policy` -> `update_read_capabilities` for every index and matview, and also iterates through all catalog objects.
- `ActiveInstance::update_write_frontiers` -> `update_read_capabilities` does really complex stuff for every Compute collection.

Reply:
Cool - if none of you are worried about blocking the coordinator thread, then neither am I! Updated the implementation to use `bootstrap_dataflow_as_of`, imitating how it's called when bootstrapping an MV.